device/sycl/src/clusterization/clusters_sum.hpp

```diff
@@ -0,0 +1,31 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s).
+#include "traccc/clusterization/detail/sparse_ccl.hpp"
+#include "traccc/edm/cell.hpp"
+
+// Vecmem include(s).
+#include <vecmem/containers/data/jagged_vector_view.hpp>
+#include <vecmem/memory/memory_resource.hpp>
+#include <vecmem/memory/unique_ptr.hpp>
+
+namespace traccc::sycl {
+
+/// Forward declaration of clusters sum function
+///
+void clusters_sum(
+    const cell_container_types::host& cells_per_event,
+    vecmem::data::jagged_vector_view<unsigned int> sparse_ccl_indices_view,
+    vecmem::unique_alloc_ptr<unsigned int>& total_clusters,
```

Comment:
This is not a good way for passing this variable to the function. :thinking: A general rule of thumb is that unique pointers should only ever be passed around by value. When you pass a unique pointer by (const) reference, you should instead just pass the underlying object by (const) reference. Since you don't want to pass ownership to the function. And at that point the function shouldn't have to know how the lifetime of the object in question is managed in the caller site.

Long story short, just pass this as `unsigned int&`. :wink:

Reply:
Hm, I see. So I actually need the pointer inside this kernel so I will just pass the raw pointer then.

Reply:
Inside the function, when starting the kernel, yes. You should pass a pointer there.

But in the signature of this function you should have a reference. That way the caller of this function is forced to provide a valid object. (Or at least they have to go out of their way to pass a non-valid reference.) Whenever a function receives a pointer, you have to add a check that it would be a valid pointer. So generally we only pass things by pointer in modern C\+\+ when the argument is allowed to be a null pointer as well. (For optional arguments in general.)

Reply:
I don't believe I follow... (I understand why we don't usually pass raw pointers in modern C++) but here, I need this pointer for the atomic addition inside the kernel. And the results of this addition I need back in the main clusterization algorithm. 

So do you mean that I should pass total clusters by `unsigned int&`, then inside the clusters_sum function I create the vecmem unique ptr, pass it to the kernel, do atomics on it and then after the kernel finishes, I assign the results from the pointer to the `unsigned int& total_clusters`?

Reply:
I mean that you should implement the code as:

```c++
void clusters_sum(..., unsigned int& total_clusters, ...) {
   ...
   queue.parallel_for<...>(range, [..., total_clusters = &total_clusters, ...] {...});
   ...
}
```

I.e. the function should receive a const reference. But you will need to pass a pointer to the kernel inside the function. Because non-const references can't easily be passed to a kernel. (If at all...)

Reply:
Gotch ya, done.